File: backend/src/SmartPayEngine.Core/Entities/ReversalRequest.cs
using SmartPayEngine.Core.Common;
using SmartPayEngine.Core.Enums;
using SmartPayEngine.Core.Exceptions;
using SmartPayEngine.Core.Models;
using SmartPayEngine.Core.ValueObjects;

namespace SmartPayEngine.Core.Entities;

/// <summary>
/// A request to reverse all or part of a transaction. Carries its own
/// probability/recommendation so it can be queued and ranked. Invariants:
/// the requested amount must be positive and not exceed the original; a full
/// reversal must equal the original amount.
/// </summary>
public sealed class ReversalRequest
{
    public Guid Id { get; }

    public Guid TransactionId { get; }

    public Money RequestedAmount { get; }

    public ReversalStatus Status { get; private set; }

    /// <summary>Engine-estimated success probability in [0, 100].</summary>
    public int SuccessProbability { get; private set; }

    public RecommendedAction RecommendedAction { get; private set; }

    public ReasonCode ReasonCode { get; private set; }

    public IReadOnlyList<EvidenceType> EvidenceNeeded { get; private set; }

    public string AIExplanation { get; private set; }

    public DateTimeOffset CreatedAt { get; }

    private ReversalRequest(
        Guid id,
        Guid transactionId,
        Money requestedAmount,
        ReversalStatus status,
        int successProbability,
        RecommendedAction recommendedAction,
        ReasonCode reasonCode,
        IReadOnlyList<EvidenceType> evidenceNeeded,
        string aIExplanation,
        DateTimeOffset createdAt)
    {
        Id = id;
        TransactionId = transactionId;
        RequestedAmount = requestedAmount;
        Status = status;
        SuccessProbability = successProbability;
        RecommendedAction = recommendedAction;
        ReasonCode = reasonCode;
        EvidenceNeeded = evidenceNeeded;
        AIExplanation = aIExplanation;
        CreatedAt = createdAt;
    }

    /// <summary>
    /// Create and validate a reversal request against the original transaction
    /// amount.
    /// </summary>
    /// <exception cref="DomainException">When an invariant is violated.</exception>
    public static ReversalRequest Create(
        Guid transactionId,
        Money requestedAmount,
        Money originalAmount,
        ReasonCode reasonCode,
        ReversalStatus status = ReversalStatus.Draft,
        DateTimeOffset? createdAt = null,
        Guid? id = null)
    {
        if (requestedAmount.Currency != originalAmount.Currency)
        {
            throw new DomainException(
                "Reversal currency must match the original transaction.",
                nameof(requestedAmount));
        }

        if (requestedAmount.IsZero)
        {
            throw new DomainException("Reversal amount must be greater than zero.", nameof(requestedAmount));
        }

        if (requestedAmount > originalAmount)
        {
            throw new DomainException(
                "Reversal amount cannot exceed the original amount.",
                nameof(requestedAmount));
        }

        return new ReversalRequest(
            id ?? Guid.NewGuid(),
            transactionId,
            requestedAmount,
            status,
            successProbability: 0,
            recommendedAction: RecommendedAction.ManualReview,
            reasonCode,
            evidenceNeeded: Array.Empty<EvidenceType>(),
            aIExplanation: string.Empty,
            createdAt ?? DateTimeOffset.UtcNow);
    }

    /// <summary>
    /// Build a request directly from an engine <see cref="ReversalAnalysis"/>,
    /// copying over the probability, action, reason code and evidence.
    /// </summary>
    public static ReversalRequest FromAnalysis(
        ReversalAnalysis analysis,
        Money originalAmount,
        DateTimeOffset? createdAt = null,
        Guid? id = null)
    {
        var reason = analysis.BestReasonCode ?? ReasonCode.ProcessingError;
        var amount = analysis.RecommendedAmount.IsZero
            ? originalAmount
            : analysis.RecommendedAmount;

        var request = Create(
            analysis.TransactionId,
            amount,
            originalAmount,
            reason,
            ReversalStatus.Submitted,
            createdAt,
            id);

        request.ApplyAnalysis(analysis);
        return request;
    }

    /// <summary>Overlay engine output onto an existing request.</summary>
    public void ApplyAnalysis(ReversalAnalysis analysis)
    {
        SuccessProbability = Scoring.ClampPercent(analysis.SuccessProbability);
        RecommendedAction = analysis.RecommendedAction;
        if (analysis.BestReasonCode is { } code)
        {
            ReasonCode = code;
        }

        EvidenceNeeded = analysis.RequiredEvidence;
        AIExplanation = analysis.AIExplanation;
    }

    /// <summary>Fraction of the original amount being reclaimed, [0, 1].</summary>
    public decimal ReclaimRatio(Money originalAmount) => RequestedAmount.RatioTo(originalAmount);

    /// <summary>
    /// Priority in [0, 100], blending success probability (winnability) with
    /// absolute exposure (size) so big and likely cases bubble up.
    /// </summary>
    public int PriorityScore
    {
        get
        {
            var probability = SuccessProbability / 100.0;
            var exposure = Scoring.Clamp((double)RequestedAmount.Amount / 1000.0, 0, 1);
            return (int)Math.Round((probability * 0.7 + exposure * 0.3) * 100);
        }
    }

    public void Submit() => Status = ReversalStatus.Submitted;

    public void Approve() => Status = ReversalStatus.Approved;

    public void Reject() => Status = ReversalStatus.Rejected;
}

File: src/domain/entities/reversal-request.ts
/**
 * `ReversalRequest` entity.
 *
 * Represents a request (manual or engine-recommended) to reverse all or part
 * of a transaction. It carries its own probability score and priority so it
 * can be queued and ranked. The entity enforces the core invariants:
 *   - a partial reversal must be > 0 and < the original amount;
 *   - a full reversal must equal the original amount;
 *   - currency must match the transaction.
 */
import { Money } from "../money";
import { DomainErrorCode, err, ok, type Result } from "../shared/result";
import { clamp } from "../shared/math";
import { EvidenceType, ReversalReasonCode } from "./reason-code";

/** Whether the request seeks the entire amount or a portion. */
export enum ReversalType {
  Full = "full",
  Partial = "partial",
  /** The engine recommends NOT pursuing a reversal. */
  None = "none",
}

/** Workflow status — aligned with the app's existing reversals UI. */
export enum ReversalStatus {
  Draft = "draft",
  Submitted = "submitted",
  UnderReview = "under_review",
  Approved = "approved",
  PartiallyApproved = "partially_approved",
  Rejected = "rejected",
}

/** A piece of evidence attached to the case. */
export interface EvidenceItem {
  readonly type: EvidenceType;
  readonly name: string;
  readonly attachedAt: Date;
}

export interface ReversalRequestProps {
  readonly id: string;
  readonly transactionId: string;
  readonly type: ReversalType;
  readonly reasonCode: ReversalReasonCode;
  /** Amount sought. For a full reversal this equals the original amount. */
  readonly amount: Money;
  /** The original transaction amount, used to validate partial bounds. */
  readonly originalAmount: Money;
  readonly status: ReversalStatus;
  readonly evidence: readonly EvidenceItem[];
  /** Success probability in [0, 100]; defaults to 0 until scored. */
  readonly successProbability?: number;
  readonly createdAt: Date;
}

export class ReversalRequest {
  private constructor(private props: ReversalRequestProps) {}

  static create(props: ReversalRequestProps): Result<ReversalRequest> {
    if (!props.transactionId.trim()) {
      return err(DomainErrorCode.Validation, "transactionId is required", "transactionId");
    }
    if (props.amount.currency !== props.originalAmount.currency) {
      return err(
        DomainErrorCode.Invariant,
        "Reversal currency must match the original transaction",
        "amount.currency",
      );
    }
    if (props.type === ReversalType.None) {
      // A "none" recommendation should carry no amount.
      if (!props.amount.isZero) {
        return err(
          DomainErrorCode.Validation,
          "A 'none' reversal must have a zero amount",
          "amount",
        );
      }
      return ok(new ReversalRequest(props));
    }
    if (!props.amount.isPositive) {
      return err(DomainErrorCode.Validation, "Reversal amount must be greater than zero", "amount");
    }
    if (props.amount.minor > props.originalAmount.minor) {
      return err(
        DomainErrorCode.Validation,
        "Reversal amount cannot exceed the original amount",
        "amount",
      );
    }
    if (props.type === ReversalType.Full && props.amount.minor !== props.originalAmount.minor) {
      return err(
        DomainErrorCode.Validation,
        "A full reversal must equal the original amount",
        "amount",
      );
    }
    if (props.type === ReversalType.Partial && props.amount.minor >= props.originalAmount.minor) {
      return err(
        DomainErrorCode.Validation,
        "A partial reversal must be strictly less than the original amount",
        "amount",
      );
    }
    return ok(new ReversalRequest(props));
  }

  get id(): string {
    return this.props.id;
  }
  get transactionId(): string {
    return this.props.transactionId;
  }
  get type(): ReversalType {
    return this.props.type;
  }
  get reasonCode(): ReversalReasonCode {
    return this.props.reasonCode;
  }
  get amount(): Money {
    return this.props.amount;
  }
  get originalAmount(): Money {
    return this.props.originalAmount;
  }
  get status(): ReversalStatus {
    return this.props.status;
  }
  get evidence(): readonly EvidenceItem[] {
    return this.props.evidence;
  }
  get successProbability(): number {
    return clamp(this.props.successProbability ?? 0, 0, 100);
  }
  get createdAt(): Date {
    return this.props.createdAt;
  }

  /** Fraction of the original amount being reclaimed, in [0, 1]. */
  get reclaimRatio(): number {
    return this.props.amount.ratioTo(this.props.originalAmount);
  }

  /** Does the case already carry the given evidence type? */
  hasEvidence(type: EvidenceType): boolean {
    return this.props.evidence.some((e) => e.type === type);
  }

  /**
   * Priority score in [0, 100] used to rank the work queue. It blends the
   * success probability (we want winnable cases first) with the absolute
   * exposure (bigger money first), so a near-certain small case and a coin-flip
   * large case are weighed sensibly.
   */
  get priorityScore(): number {
    const prob = this.successProbability / 100; // [0,1]
    // Diminishing-returns weighting of amount: ~$1k saturates toward 1.
    const exposure = clamp(this.props.amount.minor / 100_000, 0, 1);
    return Math.round((prob * 0.7 + exposure * 0.3) * 100);
  }

  /** Return a copy with an updated probability score (immutability helper). */
  withProbability(probability: number): ReversalRequest {
    return new ReversalRequest({
      ...this.props,
      successProbability: clamp(probability, 0, 100),
    });
  }
}
